From: Hannes Reinecke <hare@suse.de>
Date: Fri, 20 Apr 2018 15:44:40 +0200
Subject: [PATCH] sd: disable logical block provisioning if 'lbpme' is not set
Patch-Mainline: submitted linux-scsi 2018/10/03
References: bsc#1086095 bsc#1078355

When evaluating the 'block limits' VPD page we need to check if
the 'lbpme' (logical block provisioning managment enable) bit
is set in the READ CAPACITY (16) output.
If it isn't we can safely assume that we cannot use DISCARD on
this device.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/sd.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -415,6 +415,13 @@ provisioning_mode_store(struct device *d
 	if (sdp->type != TYPE_DISK)
 		return -EINVAL;
 
+	/*
+	 * If logical block provisioning isn't enabled we can only
+	 * select 'disable' here.
+	 */
+	if (!sdkp->lbpme && strncmp(buf, lbp_mode[SD_LBP_DISABLE], 20))
+		return -EINVAL;
+
 	if (!strncmp(buf, lbp_mode[SD_LBP_UNMAP], 20))
 		sd_config_discard(sdkp, SD_LBP_UNMAP);
 	else if (!strncmp(buf, lbp_mode[SD_LBP_WS16], 20))
@@ -2789,8 +2796,10 @@ static void sd_read_block_limits(struct
 
 		sdkp->max_ws_blocks = (u32)get_unaligned_be64(&buffer[36]);
 
-		if (!sdkp->lbpme)
+		if (!sdkp->lbpme) {
+			sd_config_discard(sdkp, SD_LBP_DISABLE);
 			goto out;
+		}
 
 		lba_count = get_unaligned_be32(&buffer[20]);
 		desc_count = get_unaligned_be32(&buffer[24]);
